common/src/sphere_collider.rs

```diff
@@ -0,0 +1,294 @@
+                solve_sphere_line_intersection(ctx.ray, &edge_pos, &edge_dir, self.radius.cosh());
+
+            // If tanh_distance is out of range or NaN, no collision occurred.
+            if !(tanh_distance >= 0.0 && tanh_distance < result.tanh_distance) {
+                continue;
+            }
+
+            let ray_endpoint = ctx.ray.ray_point(tanh_distance);
+            let contact_point = -edge_pos * math::mip(&ray_endpoint, &edge_pos)
+                + edge_dir * math::mip(&ray_endpoint, &edge_dir);
+
+            // Compute the t-coordinate of the voxels at the contact point
+            let Some(voxel_t) = dual_to_voxel(ctx, contact_point[t_axis] / contact_point.w) else {
+                continue;
+            };
+
+            // Ensure that the edge has a solid voxel adjacent to it
+            if (0..2).all(|du| {
+                (0..2).all(|dv| {
+                    ctx.get_voxel(tuv_to_xyz(t_axis, [voxel_t, u + du, v + dv])) == Material::Void
+                })
+            }) {
+                continue;
+            }
+
+            // A collision was found. Update the result.
+            result.update(ctx, tanh_distance, ray_endpoint - contact_point);
+        }
+    }
+
+    /// Detect collisions where a sphere contacts a voxel vertex
+    fn find_vertex_collision(&self, ctx: &ChunkRayTracingContext, result: &mut RayTracingResult) {
+        let float_dimension = ctx.dimension as f32;
+
+        // Loop through all grid points contained in the bounding box
+        for (x, y, z) in ctx.bounding_box.grid_point_iterator(0, 1, 2) {
+            // Skip vertices that have no solid voxels adjacent to them
+            if (0..2).all(|dx| {
+                (0..2).all(|dy| {
+                    (0..2).all(|dz| ctx.get_voxel([x + dx, y + dy, z + dz]) == Material::Void)
+                })
+            }) {
+                continue;
+            }
+
+            // Determine the cube-centric coordinates of the vertex
+            let vertex_position = math::lorentz_normalize(
+                &na::Vector3::new(x as f32, y as f32, z as f32)
+                    .scale(Vertex::chunk_to_dual_factor() as f32 / float_dimension)
+                    .insert_row(3, 1.0),
+            );
+
+            let tanh_distance =
+                solve_sphere_point_intersection(ctx.ray, &vertex_position, self.radius.cosh());
+
+            // If tanh_distance is out of range or NaN, no collision occurred.
+            if !(tanh_distance >= 0.0 && tanh_distance < result.tanh_distance) {
+                continue;
+            }
+
+            // A collision was found. Update the result.
+            let ray_endpoint = ctx.ray.ray_point(tanh_distance);
+            result.update(ctx, tanh_distance, ray_endpoint - vertex_position);
+        }
+    }
+}
+
+/// Finds the tanh of the distance a sphere will have to travel along a ray before it
+/// intersects the given point.
+///
+/// Returns NaN if there's no such intersection
```

Comment:
It's very easy to handle possibly-`NaN` data incorrectly. Should this and below to return `Option<f32>` instead?

Reply:
Possibly. It would make the code calling it slightly easier to read, and any performance cost should be minor compared to the vector arithmetic required.